### ISSUE TITLE
OOM with javascript + WebAssembly by Gramine in SGX mode

### ISSUE THREAD

**Issue by Glenrun** [gramineproject/gramine#927]
## Description of the problem
I'm running javascript + WebAssembly  in Gramine in SGX mode, but run into OOM error, however I'm able to run javascript + WebAssembly program successfully without Gramine.

## Steps to reproduce
The attachement is a complete reproducer, in dockerfile format(`gramineproject/gramine:v1.2` as base image
), however the problem has nothing to do with container, since I can reproduce this issue in bare metal host, but only because dockerfile can make us have a consistent environment.

To reproduce:
1. build docker image by `docker build -t gramine-js .`
2. run as a daemon container: `docker run -d --device /dev/sgx_enclave --device /dev/sgx_provision gramine-js:latest`
3. login container interactively: `docker exec -ti ${container_name} bash`
4. run `./startup.sh` to start `javascript + WebAssembly + SGX + Gramine` from within container instance, you should be able to gramine debug log, you could see `Out of Memory` error in between
5. run `javascript + WebAssembly` program without `Gramine and SGX` from within the same container instance: ` node index.js`, you can also invoke the server by: `curl http://localhost:3000/genAndVerify`, it should log some successful messages on node server side.

## Expected results
server should run and listen on port 3000

## Actual results
server reports `Out of memory` error

## Gramine commit hash
commit: 1.2

## Reproducer
[reproducer.tar.gz](https://github.com/gramineproject/gramine/files/9622784/reproducer.tar.gz)

**Comment by dimakuv** [comment]
I suggest two things:

1. Check https://github.com/gramineproject/gramine/issues/874 and try the solutions proposed there: increase `sgx.enclave_size` in your manifest file (I see that currently it is 8GB, I would recommend to go all the way up to 64GB) and try the `--wasm-max-mem-pages` command line option.
2. I suggest to set `-Dlog_level=all` (in your `startup.sh` script), so that Gramine prints out **all** information about what's happenning with the app. This may help root cause the problem. Please try to analyze the resulting log a bit and feel free to post it (or better the relevant part of it) here.

**Comment by Glenrun** [comment]
> I suggest two things:
> 
> 1. Check [WebAssembly can not allocate memory properly #874](https://github.com/gramineproject/gramine/issues/874) and try the solutions proposed there: increase `sgx.enclave_size` in your manifest file (I see that currently it is 8GB, I would recommend to go all the way up to 64GB) and try the `--wasm-max-mem-pages` command line option.

> 2. I suggest to set `-Dlog_level=all` (in your `startup.sh` script), so that Gramine prints out **all** information about what's happenning with the app. This may help root cause the problem. Please try to analyze the resulting log a bit and feel free to post it (or better the relevant part of it) here.

Hi, Dimitrii, I know the issue  [WebAssembly can not allocate memory properly #874](https://github.com/gramineproject/gramine/issues/874) , actually we are working on the same project, since that issue was closed due to `repackaged version of Gramine`, so I used `gramineproject/gramine:v1.2` as the base image, just add my own program, and configure `sgx.enclave_size="8G"` just for easy reproduce, for other size larger than 16G, program could be booted, but if the node receives any request from client, it reports `"can not allocate memory"` issue.

**Comment by boryspoplawski** [comment]
Could you add `log_level = "all"` to the manifest file and attach the result log here?

**Comment by Glenrun** [comment]
> Could you add `log_level = "all"` to the manifest file and attach the result log here?

Here are gramine log and application log files:
[gramine.log](https://github.com/gramineproject/gramine/files/9626517/gramine.log)
[nodejs.log](https://github.com/gramineproject/gramine/files/9626546/nodejs.log)

**Comment by dimakuv** [comment]
Gramine log has this problematic point:

```
[P1:T1:node] trace: ---- shim_mmap(0, 0x28000f000, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_NORESERVE, -1, 0x0) ...
[P1:T1:node] trace: ---- return from shim_mmap(...) = -12
```

So your NodeJS/WebAssembly app for some reason wants to allocate `0x28000f000` which is ~10GB. Clearly that's the root cause. Now why allocating this memory region, and why so much space?

I couldn't find any explanation why your app (or the underlying frameworks) do this on the internet. Can you enable more logs in Node.js and WEbAssembly? We need to understand why the app/framework tries to do this.

@boryspoplawski Actually, can't we only do LibOS memory bookkeeping for `PROT_NONE` + `MAP_NORESERVE` memory regions and mark them in VMAs as "not committed"? And then when we have an overlapping `mmap()`, we do the actual PAL-level memory commitment? (I understand that this is a pretty complex modification to Gramine, but this looks like something meaningful to do anyway.)

**Comment by Glenrun** [comment]
> Gramine log has this problematic point:
> 
> ```
> [P1:T1:node] trace: ---- shim_mmap(0, 0x28000f000, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_NORESERVE, -1, 0x0) ...
> [P1:T1:node] trace: ---- return from shim_mmap(...) = -12
> ```
> 
> So your NodeJS/WebAssembly app for some reason wants to allocate `0x28000f000` which is ~10GB. Clearly that's the root cause. Now why allocating this memory region, and why so much space?
> 
> I couldn't find any explanation why your app (or the underlying frameworks) do this on the internet. Can you enable more logs in Node.js and WEbAssembly? We need to understand why the app/framework tries to do this.
> 
The application could be modified to a very simple, just a print, after I tested, seems it is caused by WebAssembly, it allocates very large virtual memory, not physical memory, it is a known behavior:
[Chasing Memory Bugs through V8 and WebAssembly](https://bugs.chromium.org/p/v8/issues/detail?id=12076)
[Issue 12076: Hard limit on number of WebAssembly Memories](https://blog.stackblitz.com/posts/debugging-v8-webassembly/)

Referrence:
The short version is "because security". This 10GiB of address space is not actual virtual memory that V8 is requesting to the kernel. It is just an address range they will use to map "virtual" addresses in the given Memory to actual pointers in the underlying machine. By reserving such a large space, they can "sandwich" said pointers between two large guard regions of unusable memory. 


> @boryspoplawski Actually, can't we only do LibOS memory bookkeeping for `PROT_NONE` + `MAP_NORESERVE` memory regions and mark them in VMAs as "not committed"? And then when we have an overlapping `mmap()`, we do the actual PAL-level memory commitment? (I understand that this is a pretty complex modification to Gramine, but this looks like something meaningful to do anyway.)

**Comment by boryspoplawski** [comment]
> @boryspoplawski Actually, can't we only do LibOS memory bookkeeping for PROT_NONE + MAP_NORESERVE memory regions and mark them in VMAs as "not committed"? And then when we have an overlapping mmap(), we do the actual PAL-level memory commitment? (I understand that this is a pretty complex modification to Gramine, but this looks like something meaningful to do anyway.)

But what's the point? This memory would still need to be there for SGX1 and for SGX2 doing this would be a security vuln.

**Comment by Glenrun** [comment]
Case 2: 
set enclave_size=16G, Gramine can be booted, but if a request sent to server by `curl http://localhost:3000/genAndVerify` , seems even `2G` memory can't be allocated, it fails with below error:

```
Gramine is starting. Parsing TOML manifest file, this may take some time...
-----------------------------------------------------------------------------------------------------------------------
Gramine detected the following insecure configurations:

  - loader.log_level = warning|debug|trace|all (verbose log level, may leak information)
  - loader.insecure__use_cmdline_argv = true   (forwarding command-line args from untrusted host to the app)
  - loader.insecure__use_host_env = true       (forwarding environment vars from untrusted host to the app)
  - sys.insecure__allow_eventfd = true         (host-based eventfd is enabled)
  - sgx.allowed_files = [ ... ]                (some files are passed through from untrusted host without verification)

Gramine will continue application execution, but this configuration must not be used in production!
-----------------------------------------------------------------------------------------------------------------------

Example app listening on port 3000
Could not allocate 2147418112 bytes. This may cause severe instability. Trying with 1073709056 bytes
Could not allocate 1073676288 bytes. This may cause severe instability. Trying with 536838144 bytes
Could not allocate 536805376 bytes. This may cause severe instability. Trying with 268402688 bytes
Could not allocate 268369920 bytes. This may cause severe instability. Trying with 134184960 bytes
Could not allocate 134152192 bytes. This may cause severe instability. Trying with 67076096 bytes
Could not allocate 67043328 bytes. This may cause severe instability. Trying with 33521664 bytes
Could not allocate 33488896 bytes. This may cause severe instability. Trying with 16744448 bytes
Could not allocate 16711680 bytes. This may cause severe instability. Trying with 8355840 bytes
Could not allocate 8323072 bytes. This may cause severe instability. Trying with 4161536 bytes
Could not allocate 4128768 bytes. This may cause severe instability. Trying with 2064384 bytes
Could not allocate 2031616 bytes. This may cause severe instability. Trying with 1015808 bytes
Could not allocate 983040 bytes. This may cause severe instability. Trying with 491520 bytes
Could not allocate 458752 bytes. This may cause severe instability. Trying with 229376 bytes
Could not allocate 196608 bytes. This may cause severe instability. Trying with 98304 bytes
/app/node_modules/circom_runtime/build/main.cjs:85
            memory = new WebAssembly.Memory({initial:memorySize});
                     ^

RangeError: WebAssembly.Memory(): could not allocate memory
    at Object.builder [as WitnessCalculatorBuilder] (/app/node_modules/circom_runtime/build/main.cjs:85:22)
    at wtnsCalculate (/app/node_modules/snarkjs/build/main.cjs:1131:37)
    at async Object.groth16FullProve [as fullProve] (/app/node_modules/snarkjs/build/main.cjs:1175:5)
    at async run (/app/index.js:27:38)
    at async /app/index.js:15:14

Node.js v18.9.0
```

**Comment by dimakuv** [comment]
> But what's the point? This memory would still need to be there for SGX1...

@boryspoplawski What do you mean by "would still need to be there"? Sure, this memory will need to be specified in `sgx.enclave_size`. But it doesn't need to be measured at boot time and it can be skipped for zeroing-out via `sgx.preheat_enclave = false`. So this enclave memory won't be ever committed to any host physical memory.

Or am I missing something?

> ... and for SGX2 doing this would be a security vuln.

Why?

@Glenrun Unfortunately, I don't see any good workaround for your problem right now :(

**Comment by boryspoplawski** [comment]
> What do you mean by "would still need to be there"? Sure, this memory will need to be specified in sgx.enclave_size. But it doesn't need to be measured at boot time and it can be skipped for zeroing-out via sgx.preheat_enclave = false. So this enclave memory won't be ever committed to any host physical memory.

We would have EPC for it upfront. Sure it can be later swapped to host RAM to be never reclaimed, but it would need to be allocated at the beginning.
Side note: `sgx.preheat_enclave` does not zero-out memory.

>>    ... and for SGX2 doing this would be a security vuln.
>>
> Why?

I don't remember now what scenario I had in mind, maybe it's all ok.

**Comment by dimakuv** [comment]
Ok, yes, I agree with the EPC point. I admit my workaround is more of a hack. The real solution is to have EDMM.

> Side note: `sgx.preheat_enclave` does not zero-out memory.

Yes, bad wording. Instead of "zero-out memory", I should have said "touches memory".

**Comment by mccoysc** [comment]
> Ok, yes, I agree with the EPC point. I admit my workaround is more of a hack. The real solution is to have EDMM.
> 
> 
> 
> > Side note: `sgx.preheat_enclave` does not zero-out memory.
> 
> 
> 
> Yes, bad wording. Instead of "zero-out memory", I should have said "touches memory".

may be encrypted file map can be applied for these huge memory alloc?

**Comment by mccoysc** [comment]
and,the file can be sparse which no disk space required for not used range

**Comment by boryspoplawski** [comment]
> may be encrypted file map can be applied for these huge memory alloc?

I don't understand your proposal, but I'm fairly certain it's not possible in Gramine

**Comment by mccoysc** [comment]
in implemention of mmap,you can set a threshold (512MB for example).if memory map size is lower than the threshold,you can map real memory。
for the other end,if the map size is greater than the threshold,you need create a file map(the backend file should be sparse so that it could not requires disk space immediately) instead of map real memory to avoid alloc memory with huge size and return the file map address.

**Comment by boryspoplawski** [comment]
No, that's not possible. Mapping a file inside enclave is essentially mapping anonymous memory and copying file content into it. Since this is enclave memory, it must be reflected in enclave_size and backed by EPC pages (in SGX1), there is no way around that.

**Comment by mkow** [comment]
> enclave size just mean physical memory,not the address space

No, it's exactly the opposite. Enclave size is the size of `ELRANGE` in SGX terminology, which is the virtual address range reserved for a specific enclave.

**Comment by mccoysc** [comment]
> > enclave size just mean physical memory,not the address space
> 
> 
> 
> No, it's exactly the opposite. Enclave size is the size of `ELRANGE` in SGX terminology, which is the virtual address range reserved for a specific enclave.

it seems that gramine-sgx need to have a way to unbindibg the 1 to 1 relation-ship between address space and physical memory,edmm for example.

and then,the enclave size would changed to be only "address range size limit".

**Comment by weijiguo** [comment]
> after we have EDMM support
Has EDMM support been scheduled or planned?

**Comment by dimakuv** [comment]
> What we technically could implement after we have EDMM support is lazy mmaping of super-large files, i.e. internally mark a memory range as file-mapped and then load pages on-demand after a pagefault. But the implementation would be complex, as we can't just map the files 1-1 from the host, we need encryption and/or integrity checking, depending on mount type. And exceptions are super-expensive in SGX... And much more, it's basically a long design discussion and the answer may be that it's too complex and slow to be worth it.

Yes, seems like this is what @mccoysc implied in his proposal. But I fully agree with Michal -- while technically possible, it requires a lot of discussions on the rationale and design of this. Plus a lot of time to actually implement it and fix all the bugs.

> Has EDMM support been scheduled or planned?

EDMM support will be available in Linux 6.0 -- this Linux version will be released in October 2022 (so, very soon).

Gramine will support EDMM by the end of 2022, *hopefully*. The work is partially done.

So to use Gramine with EDMM, you'll need three pieces:
1. Hardware with Intel SGX supporting EDMM (all recent hardware)
2. Linux 6.0 or above (will be available soon)
3. Gramine with EDMM support (will be available in 1-4 months)

**Comment by mccoysc** [comment]
edmm support depends on the sgx driver…so if oot driver used,no need to wait for linux 6.0

**Comment by dimakuv** [comment]
> edmm support depends on the sgx driver…so if oot driver used,no need to wait for linux 6.0

That's also true. However, I am not aware of anyone (I mean, in production) using the OOT SGX driver these days.

**Comment by mccoysc** [comment]
> That's also true. However, I am not aware of anyone (I mean, in production) using the OOT SGX driver these days.

that's for the building procedure of sgx oot driver makes people heart broken🤣

**Comment by mkow** [comment]
> edmm support depends on the sgx driver…so if oot driver used,no need to wait for linux 6.0

It implements a different API than the driver in Linux 6.0, and it's deprecated, so we don't plan to implement it. We'll probably be slowly dropping support for OOT driver over time.

**Comment by mccoysc** [comment]
usually,if linux 6.0 released at the end of 2022,it needs 2 years or more to be used as the first choice.
if as you said,only supported in linux 6.0,dropping support for oot,would you make gramine-sgx to be a minority ？

**Comment by mkow** [comment]
No, 6.0 is only about EDMM support. Older kernels already have SGX drivers integrated and Gramine works and will work there. For even older ones, people use DCAP drivers if they have CPUs with FLC. OOT is the option only for very old CPUs without FLC.

**Comment by dimakuv** [comment]
@mkow I think the point of @mccoysc was that many people will have this:
1. Linux version less than 6.0
2. Desire to use EDMM

To satisfy the above, the only option is to install the OOT SGX driver. I agree that deprecating the OOT driver is too early -- many people will have the above (conflicting) requirements.

**Comment by mkow** [comment]
The support for OOT and the support for EDMM-on-OOT are two completely different things, and you seems to be mixing both of them. AFAIK we don't plan to ever support EDMM-on-OOT.

**Comment by mccoysc** [comment]
even,we mainly run linux with version<5.0.
too many things need to be handled but not so much benifits.

**Comment by lejunzhu** [comment]
> To satisfy the above, the only option is to install the OOT SGX driver.

AFAIK, OOT driver doesn't support DCAP, so that probably won't be useful.

**Comment by mccoysc** [comment]
really,implemention of mmap flag MAP_NORESERVE is more important for many data analyzing service

**Comment by mkow** [comment]
> even,we mainly run linux with version<5.0.

Well, then update it. You'll need to do it someday anyways ;)

> really,implemention of mmap flag MAP_NORESERVE is more important for many data analyzing service

Again, sorry, but it seems that you don't understand how SGX works. We can't just map memory as `MAP_NORESERVE` and be done, enclave memory is managed through special CPU instructions and is completely different than normal, OS-managed RAM.

**Comment by mccoysc** [comment]
not "can not",but "no enough power"

**Comment by mkow** [comment]
?

**Comment by boryspoplawski** [comment]
@mccoysc It's impossible. If you think otherwise, feel free to prove me wrong and submit a PR implementing it, you're more than welcome to do so.

**Comment by boryspoplawski** [comment]
@Glenrun for the original issue we cannot do anything, until we have SGX2 support.
For "Case 2" it's hard to say, doesn't seem like Gramine issue. You can add `log_level = "all"` and check all `mmap` accesses, maybe node already allocated memory which filled whole enclave... Hard to say without more details.

**Comment by Glenrun** [comment]
OOT driver has two types: 
[1.  Driver supports for CPU without FLC](https://github.com/intel/linux-sgx-driver)
[2. Driver supports DCAP](https://github.com/intel/SGXDataCenterAttestationPrimitives/tree/master/driver/linux)
and Since Linux kernel 5.11, the later DCAP driver is up-streamed.
So available options: 
1) Keep current Linux kernel version < 5.11, any of those two OOT drivers can be installed
2) Simply upgrade Linux kernel to 5.11 without waiting for Linux 6.0
both ways can be used to support EDMM

**Comment by dimakuv** [comment]
I believed we answered the original issue (which was out of memory due to inssuficient `sgx.enclave_size`). The issue can be solved only with EDMM, which is not yet supported in Gramine, but the support will be added soon-ish.

The discussion of three different SGX drivers (OOT non-dcap, OOT dcap, and in-Linux-kernel) is off-topic. I'm closing this issue.

**Comment by Glenrun** [comment]
@dimakuv Hi, Dmitrii, Linux 6.0 was released, any plan for edmm support?

**Comment by dimakuv** [comment]
Yes, we have a plan for EDMM support, and we already have a draft of the EDMM implementation. The hope is to get this merged into Gramine in the next couple months (could be sooner, like in 1 month, could be a bit later). But surely EDMM support is one of our highest priorities currently.
